x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/hooks/use_step_define_form.ts

```diff
@@ -0,0 +1,89 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useEffect } from 'react';
+
+import { getPreviewRequestBody } from '../../../../../common';
+
+import { getDefaultStepDefineState } from '../common';
+
+import { StepDefineFormProps } from '../step_define_form';
+
+import { useAdvancedPivotEditor } from './use_advanced_pivot_editor';
+import { useAdvancedSourceEditor } from './use_advanced_source_editor';
+import { usePivotConfig } from './use_pivot_config';
+import { useSearchBar } from './use_search_bar';
+
+export type StepDefineFormHook = ReturnType<typeof useStepDefineForm>;
+
+export const useStepDefineForm = ({ overrides, onChange, searchItems }: StepDefineFormProps) => {
+  const defaults = { ...getDefaultStepDefineState(searchItems), ...overrides };
+  const { indexPattern } = searchItems;
+
+  const searchBar = useSearchBar(defaults, indexPattern);
+  const pivotConfig = usePivotConfig(defaults, indexPattern);
+
+  const previewRequest = getPreviewRequestBody(
+    indexPattern.title,
+    searchBar.state.pivotQuery,
+    pivotConfig.state.pivotGroupByArr,
+    pivotConfig.state.pivotAggsArr
+  );
+
+  // pivot config hook
+  const stringifiedPivotConfig = JSON.stringify(previewRequest.pivot, null, 2);
+  const advancedPivotEditor = useAdvancedPivotEditor(defaults, stringifiedPivotConfig);
+
+  // source config hook
+  const stringifiedSourceConfig = JSON.stringify(previewRequest.source.query, null, 2);
+  const advancedSourceEditor = useAdvancedSourceEditor(defaults, stringifiedSourceConfig);
+
+  useEffect(() => {
+    if (!advancedSourceEditor.state.isAdvancedSourceEditorEnabled) {
+      const previewRequestUpdate = getPreviewRequestBody(
+        indexPattern.title,
+        searchBar.state.pivotQuery,
+        pivotConfig.state.pivotGroupByArr,
+        pivotConfig.state.pivotAggsArr
+      );
+
+      const stringifiedSourceConfigUpdate = JSON.stringify(
+        previewRequestUpdate.source.query,
+        null,
+        2
+      );
+
+      advancedSourceEditor.actions.setAdvancedEditorSourceConfig(stringifiedSourceConfigUpdate);
+    }
+
+    onChange({
+      aggList: pivotConfig.state.aggList,
+      groupByList: pivotConfig.state.groupByList,
+      isAdvancedPivotEditorEnabled: advancedPivotEditor.state.isAdvancedPivotEditorEnabled,
+      isAdvancedSourceEditorEnabled: advancedSourceEditor.state.isAdvancedSourceEditorEnabled,
+      searchLanguage: searchBar.state.searchLanguage,
+      searchString: searchBar.state.searchString,
+      searchQuery: searchBar.state.searchQuery,
+      sourceConfigUpdated: advancedSourceEditor.state.sourceConfigUpdated,
+      valid: pivotConfig.state.valid,
+    });
+    // custom comparison
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [
+    JSON.stringify(advancedPivotEditor.state),
+    JSON.stringify(advancedSourceEditor.state),
+    JSON.stringify(pivotConfig.state),
+    JSON.stringify(searchBar.state),
```

Comment:
there should be a better way 🤔why not make each config as observable and use `combineLatest`?  also each of the config could be exposed with `distinctUntilChanged` operator.

Reply:
yes it could be improved, note it is older code, for this PR I just copied it over from the corresponding component where it was used directly in the same way previously.